src/components/Table/GridTable.test.tsx

```diff
@@ -1445,6 +1445,99 @@ describe("GridTable", () => {
     expect(api.current!.getSelectedRowIds()).toEqual([]);
   });
 
+  it("sets the selected state of the group row as expected when children are collapsed", async () => {
+    // Given a table that can apply a filter and three parents with children.
+    const rows: GridDataRow<NestedRow>[] = [
+      simpleHeader,
+      {
+        ...{ kind: "parent", id: "p1", data: { name: "parent 1" } },
+        // One parent where all children match the filter
+        children: [
+          { kind: "child", id: "p1c1", data: { name: "filter child p1c1" } },
+          { kind: "child", id: "p1c2", data: { name: "filter child p1c2" } },
+        ],
+      },
+      {
+        ...{ kind: "parent", id: "p2", data: { name: "parent 2" } },
+        // A second parent where only one child matches the filter
+        children: [
+          { kind: "child", id: "p2c1", data: { name: "filter child p2c1" } },
+          { kind: "child", id: "p2c2", data: { name: "child p2c2" } },
+        ],
+      },
+      {
+        // A third parent that matches the filter, but no children match the filter
+        ...{ kind: "parent", id: "p3", data: { name: "filter parent 3" } },
+        children: [{ kind: "child", id: "p3c1", data: { name: "child p3c1" } }],
+      },
+    ];
+    const api: MutableRefObject<GridTableApi<NestedRow> | undefined> = { current: undefined };
+    const r = await render(<TestFilterAndSelect rows={rows} api={api} />);
+
+    // When triggering all rows as selected
+    click(cellAnd(r, 0, 1, "select"));
+    // Then expect all rows should selected
+    expect(api.current!.getSelectedRowIds()).toEqual(["p3", "p3c1", "p2", "p2c2", "p2c1", "p1", "p1c2", "p1c1"]);
+
+    // When collapsing all rows
+    click(cellAnd(r, 0, 0, "collapse"));
+    // Then expect each group row to persist the selected stated
+    expect(cellAnd(r, 1, 1, "select")).toBeChecked();
+    expect(cellAnd(r, 2, 1, "select")).toBeChecked();
+    expect(cellAnd(r, 3, 1, "select")).toBeChecked();
+
+    // And all rows are still returned by the API
+    expect(api.current!.getSelectedRowIds()).toEqual(["p3", "p3c1", "p2", "p2c2", "p2c1", "p1", "p1c2", "p1c1"]);
+
+    // When applying a filter
+    type(r.filter, "filter");
+    // Then each group rows selected state should reflect the children that match the filter
+    expect(cellAnd(r, 1, 1, "select")).toBeChecked();
+    expect(cellAnd(r, 2, 1, "select")).toBeChecked();
+    expect(cellAnd(r, 3, 1, "select")).not.toBeChecked();
+    // And the API reflects the expected selected states of rows that match the filter
+    expect(api.current!.getSelectedRowIds()).toEqual(["p2", "p2c1", "p1", "p1c2", "p1c1"]);
```

Comment:
Question: Doesn't `p3` have "filter" in the name "filter parent 3", and so should show up here?

Reply:
It isn't selected though, the parent rows selection state is based on the it's matched children. No children match, so it isn't considered selected.

Reply:
Huh, yeah, I guess... Just thinking that p2 is there, even though p2 doesn't have _all_ of its children.

Do we have a case where a filtered & selected parent (both of which seem imo like "...surely this would be considered selected") shouldn't be returned b/c of no matched children? Like a specific page/etc...


Dunno, not a huge deal, just surprised.

Reply:
I think in Specs and Selections where our group rows just are roll ups, we don't actually do anything with those rows if they're selected. We just use the select box on those group rows as a way to quickly select all its children. I think that is true for most, if not all, of the grouped tables in Blueprint.
Thinking through a use case where I want to delete line items, should I be able to delete the whole group if it is selected, even if some children under it are hidden by a filter? I am not sure... I'll follow up with UX on that one to tease it out. I could play devil's advocate for either side of that argument.
This is at least a step in the right direction, imo. Thank you for questioning other use cases, though.